dace/frontend/python/parser.py

```diff
@@ -282,50 +248,99 @@ def __call__(self, *args, **kwargs):
 
         # Invoke auto-optimization as necessary
         if Config.get_bool('optimizer', 'autooptimize') or self.auto_optimize:
-            from dace.transformation.auto import auto_optimize as autoopt
-            sdfg = autoopt.auto_optimize(sdfg, self.device)
+            sdfg = self._auto_optimize(sdfg)
 
         # Compile SDFG (note: this is done after symbol inference due to shape
         # altering transformations such as Vectorization)
         binaryobj = sdfg.compile()
 
+        self._cache = (argtypes, sdfg, binaryobj)
+
         # Call SDFG
         result = binaryobj(**kwargs)
 
-        # Enforce deletion and closure of compiled program
-        del binaryobj
-
         return result
 
+    def parse(self, *compilation_args, strict=None, save=False) -> SDFG:
+        """ 
+        Try to parse a DaceProgram object and return the `dace.SDFG` object
+        that corresponds to it.
+        :param function: DaceProgram object (obtained from the ``@dace.program``
+                        decorator).
+        :param compilation_args: Various compilation arguments e.g. dtypes.
+        :param strict: Whether to apply strict transformations or not (None
+                    uses configuration-defined value). 
+        :param save: If True, saves the generated SDFG to 
+                    ``_dacegraphs/program.sdfg`` after parsing.
+        :return: The generated SDFG object.
+        """
+        # Avoid import loop
+        from dace.sdfg.analysis import scalar_to_symbol as scal2sym
+        from dace.transformation import helpers as xfh
+
+        # Obtain DaCe program as SDFG
+        sdfg = self.generate_pdp(*compilation_args, strict=strict)
+
+        # Set argument names
+        sdfg.arg_names = self.argnames
+
+        # Apply strict transformations automatically
+        if (strict == True or (strict is None and Config.get_bool(
+                'optimizer', 'automatic_strict_transformations'))):
+
+            # Promote scalars to symbols as necessary
+            promoted = scal2sym.promote_scalars_to_symbols(sdfg)
+            if Config.get_bool('debugprint') and len(promoted) > 0:
+                print('Promoted scalars {%s} to symbols.' %
+                      ', '.join(p for p in sorted(promoted)))
+
+            sdfg.apply_strict_transformations()
+
+            # Split back edges with assignments and conditions to allow richer
+            # control flow detection in code generation
+            xfh.split_interstate_edges(sdfg)
+
+        # Save the SDFG. Skip this step if running from a cached SDFG, as
+        # it might overwrite the cached SDFG.
+        if not Config.get_bool('compiler', 'use_cache') and save:
+            sdfg.save(os.path.join('_dacegraphs', 'program.sdfg'))
```

Comment:
I know this is not new, but out of curiosity -- why do we always save the SDFG here? It's already saved in the `.dacecache`.

Reply:
we now don't :) the diff that for some reason doesn't show up is that the argument is now `save=False` 💯

Reply:
Ah, gr8